In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("data_set_utf8.csv")

In [3]:
df.dtypes

id                  object
district            object
price               object
price_per_m2        object
floor               object
floor_number        object
prod_year           object
rent                object
market              object
building_type       object
finish_condition    object
flat_surface_m2     object
room_number         object
dtype: object

In [4]:
df.shape

(15036, 13)

In [5]:
print(df.isna().sum().sort_values(ascending=False))
# print(df.isna().sum())

rent                9717
finish_condition    8553
building_type       5237
prod_year           2508
floor_number        2454
floor                376
price_per_m2         296
room_number            0
flat_surface_m2        0
market                 0
price                  0
district               0
id                     0
dtype: int64


In [6]:
def missing(dff):
    print (round((dff.isnull().sum() * 100/ len(dff)),2).sort_values(ascending=False))

In [7]:
df.head(30)

id                                           district  \
0   offer-item-ad_id48H25            Mieszkanie na sprzedaż: Wrocław, Krzyki   
1   offer-item-ad_id48IAh            Mieszkanie na sprzedaż: Wrocław, Krzyki   
2   offer-item-ad_id48Ihn      Mieszkanie na sprzedaż: Wrocław, Stare Miasto   
3   offer-item-ad_id47fLE      Mieszkanie na sprzedaż: Wrocław, Stare Miasto   
4   offer-item-ad_id45AYc      Mieszkanie na sprzedaż: Wrocław, Stare Miasto   
5   offer-item-ad_id44eWl            Mieszkanie na sprzedaż: Wrocław, Krzyki   
6   offer-item-ad_id44gg6            Mieszkanie na sprzedaż: Wrocław, Krzyki   
7   offer-item-ad_id48Le4  Mieszkanie na sprzedaż: Wrocław, Krzyki, Południe   
8   offer-item-ad_id48jKt      Mieszkanie na sprzedaż: Wrocław, Stare Miasto   
9   offer-item-ad_id450d0      Mieszkanie na sprzedaż: Wrocław, Stare Miasto   
10  offer-item-ad_id45fE0      Mieszkanie na sprzedaż: Wrocław, Stare Miasto   
11  offer-item-ad_id48F8M     Mieszkanie na sprzedaż: Wrocław, Krzyki, Borek   
12  offer-item-ad_id40lxC      Mieszkanie na sprzedaż: Wrocław, Stare Miasto   
13  offer-item-ad_id44gjc            Mieszkanie na sprzedaż: Wrocław, Krzyki   
14  offer-item-ad_id45aHt      Mieszkanie na sprzedaż: Wrocław, Stare Miasto   
15  offer-item-ad_id48GNG       Mieszkanie na sprzedaż: Wrocław, Śródmieście   
16  offer-item-ad_id48DDm      Mieszkanie na sprzedaż: Wrocław, Stare Miasto   
17  offer-item-ad_id48S5e       Mieszkanie na sprzedaż: Wrocław, Śródmieście   
18  offer-item-ad_id482Vr      Mieszkanie na sprzedaż: Wrocław, Stare Miasto   
19  offer-item-ad_id48ndA  Mieszkanie na sprzedaż: Wrocław, Śródmieście, ...   
20  offer-item-ad_id45YkZ          Mieszkanie na sprzedaż: Wrocław, Nadodrze   
21  offer-item-ad_id44gl7            Mieszkanie na sprzedaż: Wrocław, Krzyki   
22  offer-item-ad_id44eVA            Mieszkanie na sprzedaż: Wrocław, Krzyki   
23  offer-item-ad_id488G4      Mieszkanie na sprzedaż: Wrocław, Stare Miasto   
24  offer-item-ad_id45HcN      Mieszkanie na sprzedaż: Wrocław, Stare Miasto   
25  offer-item-ad_id45n6D      Mieszkanie na sprzedaż: Wrocław, Stare Miasto   
26  offer-item-ad_id48S2X          Mieszkanie na sprzedaż: Wrocław, Nadodrze   
27  offer-item-ad_id48BnC  Mieszkanie na sprzedaż: Wrocław, Psie Pole, Sw...   
28  offer-item-ad_id48amt  Mieszkanie na sprzedaż: Wrocław, Fabryczna, Op...   
29  offer-item-ad_id47E0g       Mieszkanie na sprzedaż: Wrocław, Krzyki, Gaj   

           price  price_per_m2           floor       floor_number  \
0     590 000 zł   8 309 zł/m²       Piętro: 3   Liczba pięter: 3   
1     585 000 zł   9 286 zł/m²       Piętro: 3   Liczba pięter: 3   
2     459 000 zł   9 592 zł/m²       Piętro: 4   Liczba pięter: 7   
3   7 500 000 zł  35 361 zł/m²       Piętro: 8   Liczba pięter: 9   
4   7 000 000 zł  32 590 zł/m²       Piętro: 6   Liczba pięter: 6   
5   5 245 000 zł  23 106 zł/m²    Piętro: > 10  Liczba pięter: 48   
6   4 800 000 zł  27 195 zł/m²    Piętro: > 10  Liczba pięter: 48   
7   4 600 000 zł  34 074 zł/m²    Piętro: > 10  Liczba pięter: 51   
8   4 000 000 zł  29 851 zł/m²       Piętro: 1   Liczba pięter: 3   
9   3 582 800 zł  20 000 zł/m²       Piętro: 9  Liczba pięter: 10   
10  3 579 400 zł  19 889 zł/m²       Piętro: 9  Liczba pięter: 11   
11  3 500 000 zł  13 901 zł/m²  Piętro: parter   Liczba pięter: 2   
12  3 500 000 zł  29 476 zł/m²       Piętro: 6   Liczba pięter: 6   
13  3 474 555 zł  18 904 zł/m²    Piętro: > 10  Liczba pięter: 51   
14  3 470 000 zł  19 350 zł/m²       Piętro: 8  Liczba pięter: 10   
15  3 407 270 zł  19 000 zł/m²       Piętro: 8   Liczba pięter: 9   
16  3 407 270 zł  18 982 zł/m²       Piętro: 9  Liczba pięter: 10   
17  3 407 270 zł  18 929 zł/m²       Piętro: 8  Liczba pięter: 10   
18  3 400 000 zł  18 959 zł/m²       Piętro: 8  Liczba pięter: 10   
19  3 391 669 zł  11 090 zł/m²       Piętro: 6   Liczba pięter: 6   
20  3 391 669 zł  11 283 zł/m²       Piętro: 6   Liczba pięter: 6   
21  3 383 816 zł  18 904

In [8]:
# Wyświetlanie całego wiersza
# Display row's full lenght
pd.set_option('display.max_colwidth', -1)

#### Id

In [9]:
# ile pustych
# how many empty index values
len(df[df.id.isna()])
# ile unikalnych wartości
# how many unique values
len(df.id.unique())
# id
df.id.head()

0    offer-item-ad_id48H25
1    offer-item-ad_id48IAh
2    offer-item-ad_id48Ihn
3    offer-item-ad_id47fLE
4    offer-item-ad_id45AYc
Name: id, dtype: object

<b>offer-item-ad_id48H25</b> <--- wygląd pojedyńczej wartości
<br>
<b>id48H25</b> <--- wygląd docelowy
##### ENGLISH VERSION
<b>offer-item-ad_id48H25</b> <--- example of an index value
<br>
<b>id48H25</b> <--- target appearance

In [10]:
df['id'] = df['id'].apply(lambda x: x.split('_')[1])

In [11]:
df.id.head()

0    id48H25
1    id48IAh
2    id48Ihn
3    id47fLE
4    id45AYc
Name: id, dtype: object

#### district

In [12]:
df.district.head()

0    Mieszkanie na sprzedaż: Wrocław, Krzyki      
1    Mieszkanie na sprzedaż: Wrocław, Krzyki      
2    Mieszkanie na sprzedaż: Wrocław, Stare Miasto
3    Mieszkanie na sprzedaż: Wrocław, Stare Miasto
4    Mieszkanie na sprzedaż: Wrocław, Stare Miasto
Name: district, dtype: object

<b>Mieszkanie na sprzedaż: Wrocław, Fabryczna, Grabiszynek</b> <--- tak wygląda surowa wartość
<br>
<b>Fabryczna</b> <--- docelowa wartość
<br>
Dodatkowo pojawia się wartości dzielnicy:
<ul>
    <li><b>dolnośląskie</b> - w przypadku takiej wartości wstawiam <i>np.NaN</i></li>
    <li><b>Rynek</b> - w przypadku takiej wartości wstawiam <i>Stare Miasto</i></li>
    <li><b>Nadodrze</b> - w przypadku takiej wartości wstawiam <i>Śródmieście</i></li> 
</ul>

##### ENGLISH VERSION
<b>Mieszkanie na sprzedaż: Wrocław, Fabryczna, Grabiszynek</b> <--- example of a district value
<br>
<b>Fabryczna</b> <--- target appearance
<br>
Additionally there are exceptions like this: 
<ul>
    <li><b>dolnośląskie</b> - in this case, imput <i>np.NaN</i></li>
    <li><b>Rynek</b> - in this case, imput <i>Stare Miasto</i></li>
    <li><b>Nadodrze</b> - in this case, imput <i>Śródmieście</i></li> 
</ul>

In [13]:
df['district'] = df['district'].apply(lambda x: np.NaN if x.split(',')[1].strip() == 'dolnośląskie' 
                                      else x.split(',')[1].strip().replace("Rynek","Stare Miasto").replace("Nadodrze","Śródmieście")) 

In [14]:
df.district.unique()

array(['Krzyki', 'Stare Miasto', 'Śródmieście', 'Psie Pole', 'Fabryczna',
       nan], dtype=object)

In [15]:
df.district.head()

0    Krzyki      
1    Krzyki      
2    Stare Miasto
3    Stare Miasto
4    Stare Miasto
Name: district, dtype: object

#### Price

In [16]:
df.price.head()

0    590 000 zł  
1    585 000 zł  
2    459 000 zł  
3    7 500 000 zł
4    7 000 000 zł
Name: price, dtype: object

<b>590 000 zł</b> <--- tak wygląda surowa wartość
<br>
<b>590000.0</b> <--- docelowa wartość
<br>
Dodatkowo pojawia się wartości ceny: 
<ul>
    <li><b>Zapytaj o cenę</b> - w przypadku takiej wartości wstawiam <i>np.NaN</i></li>
</ul>

##### ENGLISH VERSION
<b>590 000 zł</b> <--- example of a price value
<br>
<b>590000.0</b> <--- target appearance
<br>
Additionally there are exceptions like this: 
<ul>
    <li><b>Zapytaj o cenę</b> - in this case, imput <i>np.NaN</i></li>
</ul>

In [17]:
df['price'] = df['price'].apply(lambda x: pd.to_numeric(x.split(' z')[0].strip().replace(' ', '').replace(',','.')) 
                                if x.lower() != 'zapytaj o cenę' else np.NaN)

In [18]:
df.price.head()

0    590000.0 
1    585000.0 
2    459000.0 
3    7500000.0
4    7000000.0
Name: price, dtype: float64

In [19]:
len(df.price.unique())

5697

In [20]:
len(df[df.price.isna()])

297

#### Price per m2

In [21]:
df.price_per_m2.head()

0    8 309 zł/m² 
1    9 286 zł/m² 
2    9 592 zł/m² 
3    35 361 zł/m²
4    32 590 zł/m²
Name: price_per_m2, dtype: object

<b>8 309 zł/m²</b> <--- tak wygląda surowa wartość
<br>
<b>8309.0</b> <--- docelowa wartość
##### ENGLISH VERSION
<b>8 309 zł/m²</b> <--- example of a Price per m2 value
<br>
<b>8309.0</b> <--- target appearance

In [22]:
df['price_per_m2'] = df['price_per_m2'].apply(lambda x: pd.to_numeric(x.split('z')[0].strip().replace(' ','')) 
                                              if pd.notna(x) else np.NaN)

In [23]:
df.price_per_m2.head()

0    8309.0 
1    9286.0 
2    9592.0 
3    35361.0
4    32590.0
Name: price_per_m2, dtype: float64

In [24]:
len(df[df.price_per_m2.isna()])

296

#### Floor

In [25]:
df.floor.head()

0    Piętro: 3
1    Piętro: 3
2    Piętro: 4
3    Piętro: 8
4    Piętro: 6
Name: floor, dtype: object

In [26]:
df.floor.unique()

array(['Piętro: 3', 'Piętro: 4', 'Piętro: 8', 'Piętro: 6', 'Piętro: > 10',
       'Piętro: 1', 'Piętro: 9', 'Piętro: parter', 'Piętro: 7',
       'Piętro: 2', nan, 'Piętro: 5', 'Piętro: 10', 'Piętro: poddasze',
       'Piętro: suterena'], dtype=object)

<b>Piętro: 3</b> <--- tak wygląda surowa wartość
<br>
<b>3.0</b> <--- docelowa wartość
<br>
Dodatkowo pojawia się wartości pięter takie jak: 
<ul>
    <li>
        <b>parter</b> - w przypadku takiej wartości wstawiam <i>0</i> ponieważ w polsce liczone jest to jako zerowe piętro
    </li>
    <li>
        <b>suterena</b> - w przypadku takiej wartości wstawiam <i>-1</i> ponieważ jest to mieszkanie poniżej poziomu ulicy, czasami w połowie wystające ponad
    </li>
    <li><b>poddasze</b> - w przypadku takiej wartości wstawiam <i>4</i> ponieważ postanowiłem wstawić w ty miejscu  wartość uśrednioną i zaokrągloną, liczby pięter we wszystkich wierszach posiadających wartość "poddasze" w kolumnie floor</li>
    <li>
        <b> 10</b> - w przypadku takiej wartości wstawiam<i>np.NaN</i>
    </li>
</ul>

##### ENGLISH VERSION
<b>Piętro: 3</b> <--- example of a Floor value
<br>
<b>3.0</b> <--- target appearance
<br>
Additionally there are exceptions like this: 
<ul>
    <li>
        <b>parter</b> - in this case, imput <i>0</i> because in Poland parter is not a first floor and we don't count it as a floor
    </li>
    <li>
        <b>suterena</b> - in this case, imput <i>-1</i> because it is flat in basement 
    </li>
    <li><b>poddasze</b> - in this case, imput <i>4</i> because it is mean value in evere row contains value "poddasze" in column floor</li>
    <li>
        <b> 10</b> - in this case, imput <i>np.NaN</i>
    </li>
</ul>

In [27]:
df['floor'] = df['floor'].apply(lambda x: x if x != 'Piętro: > 10' else np.NaN)

In [28]:
df['floor'] = df['floor'].apply(lambda x: pd.to_numeric(x.split(':')[1].strip().replace('parter','0').replace('suterena','-1').replace('poddasze','4')) 
                                if pd.notnull(x) else np.NaN)

In [29]:
len(df[df.floor.isna()])

456

In [30]:
df.floor.unique()

array([ 3.,  4.,  8.,  6., nan,  1.,  9.,  0.,  7.,  2.,  5., 10., -1.])

In [31]:
df.floor.head()

0    3.0
1    3.0
2    4.0
3    8.0
4    6.0
Name: floor, dtype: float64

#### Floor number

In [32]:
df.floor_number.head()

0    Liczba pięter: 3
1    Liczba pięter: 3
2    Liczba pięter: 7
3    Liczba pięter: 9
4    Liczba pięter: 6
Name: floor_number, dtype: object

In [33]:
df.floor_number.unique()

array(['Liczba pięter: 3', 'Liczba pięter: 7', 'Liczba pięter: 9',
       'Liczba pięter: 6', 'Liczba pięter: 48', 'Liczba pięter: 51',
       'Liczba pięter: 10', 'Liczba pięter: 11', 'Liczba pięter: 2',
       'Liczba pięter: 5', 'Liczba pięter: 17', nan, 'Liczba pięter: 1',
       'Liczba pięter: 8', 'Liczba pięter: 4', 'Liczba pięter: 52',
       'Liczba pięter: 50', 'Liczba pięter: 18', 'Liczba pięter: 47',
       'Liczba pięter: 16', 'Liczba pięter: 14', 'Liczba pięter: 49',
       'Liczba pięter: 15', 'Liczba pięter: 13', 'Liczba pięter: 100',
       'Liczba pięter: 12'], dtype=object)

<b>Liczba pięter: 6</b> <--- tak wygląda surowa wartość
<br>
<b>6.0</b> <--- docelowa wartość
<br>
Dodatkowo pojawia się wartoś piętra taka jak: 
<ul>
    <li>
        <b>100</b> - w przypadku takiej wartości wstawiam <i>np>NaN</i>
    </li>
</ul>

##### ENGLISH VERSION
<b>Liczba pięter: 6</b> <--- example of a Floor number value
<br>
<b>6.0</b> <--- target appearance
<br>
Additionally there are exceptions like this:
<ul>
    <li>
        <b>100</b> -in this case, impute <i>np.NaN</i>
    </li>
</ul>

In [34]:
df['floor_number'] = df['floor_number'].apply(lambda x: pd.to_numeric(x.split(':')[1].strip()) if pd.notnull(x) else np.NaN)

In [35]:
df.floor_number.head()

0    3.0
1    3.0
2    7.0
3    9.0
4    6.0
Name: floor_number, dtype: float64

In [36]:
df.floor_number.unique()

array([  3.,   7.,   9.,   6.,  48.,  51.,  10.,  11.,   2.,   5.,  17.,
        nan,   1.,   8.,   4.,  52.,  50.,  18.,  47.,  16.,  14.,  49.,
        15.,  13., 100.,  12.])

In [37]:
df.floor_number.describe()

count    12582.000000
mean     4.746702    
std      3.297912    
min      1.000000    
25%      3.000000    
50%      4.000000    
75%      6.000000    
max      100.000000  
Name: floor_number, dtype: float64

In [38]:
df.floor_number.median()

4.0

In [39]:
df[df['floor_number'] == 100]

id     district     price  price_per_m2  floor  floor_number  \
6623  id48VW2  Śródmieście  456083.0  9598.0        1.0    100.0          

             prod_year            rent            market  \
6623  Rok budowy: 2021  Czynsz: 300 zł  Rynek: pierwotny   

                         building_type finish_condition flat_surface_m2  \
6623  Rodzaj zabudowy: apartamentowiec  NaN              47,52 m²         

     room_number  
6623  3 pokoje

In [40]:
df['floor_number'] = df['floor_number'].apply(lambda x: np.NaN if x == 100 else x)

#### Production Year

In [41]:
df.prod_year.head()

0    Rok budowy: 1930
1    Rok budowy: 2008
2    Rok budowy: 2022
3    Rok budowy: 2016
4    Rok budowy: 2016
Name: prod_year, dtype: object

In [42]:
df.prod_year.unique()

array(['Rok budowy: 1930', 'Rok budowy: 2008', 'Rok budowy: 2022',
       'Rok budowy: 2016', 'Rok budowy: 2012', 'Rok budowy: 2013',
       'Rok budowy: 2021', 'Rok budowy: 2020', nan, 'Rok budowy: 2015',
       'Rok budowy: 2007', 'Rok budowy: 1990', 'Rok budowy: 2014',
       'Rok budowy: 2019', 'Rok budowy: 1911', 'Rok budowy: 1902',
       'Rok budowy: 2005', 'Rok budowy: 2018', 'Rok budowy: 1925',
       'Rok budowy: 2003', 'Rok budowy: 2010', 'Rok budowy: 1907',
       'Rok budowy: 1910', 'Rok budowy: 1935', 'Rok budowy: 1912',
       'Rok budowy: 1957', 'Rok budowy: 1900', 'Rok budowy: 2017',
       'Rok budowy: 1936', 'Rok budowy: 2009', 'Rok budowy: 2006',
       'Rok budowy: 1920', 'Rok budowy: 1903', 'Rok budowy: 1995',
       'Rok budowy: 2011', 'Rok budowy: 1906', 'Rok budowy: 2000',
       'Rok budowy: 1997', 'Rok budowy: 1932', 'Rok budowy: 2004',
       'Rok budowy: 1971', 'Rok budowy: 1975', 'Rok budowy: 1970',
       'Rok budowy: 1405', 'Rok budowy: 2002', 'Rok budow

In [43]:
print('Rok budowy: 80, ilość przypadków: ',len(df[df.prod_year == "Rok budowy: 80"]))
print('Rok budowy: 90, ilość przypadków: ',len(df[df.prod_year == "Rok budowy: 90"]))
print('Rok budowy: 93, ilość przypadków: ',len(df[df.prod_year == "Rok budowy: 93"]))
print('Rok budowy: 95, ilość przypadków: ',len(df[df.prod_year == "Rok budowy: 95"]))
print('Rok budowy: 1, ilość przypadków: ',len(df[df.prod_year == "Rok budowy: 1"]))
print('Rok budowy: 19, ilość przypadków: ',len(df[df.prod_year == "Rok budowy: 19"]))
print('Rok budowy: 202, ilość przypadków: ',len(df[df.prod_year == "Rok budowy: 202"]))
print('Rok budowy: 1196, ilość przypadków: ',len(df[df.prod_year == "Rok budowy: 1196"]))
print('Rok budowy: 20521, ilość przypadków: ',len(df[df.prod_year == "Rok budowy: 20521"]))

Rok budowy: 80, ilość przypadków:  1
Rok budowy: 90, ilość przypadków:  1
Rok budowy: 93, ilość przypadków:  1
Rok budowy: 95, ilość przypadków:  1
Rok budowy: 1, ilość przypadków:  3
Rok budowy: 19, ilość przypadków:  1
Rok budowy: 202, ilość przypadków:  1
Rok budowy: 1196, ilość przypadków:  1
Rok budowy: 20521, ilość przypadków:  1


<b>Rok budowy: 1930</b> <--- tak wygląda surowa wartość
<br>
<b>1930</b> <--- docelowa wartość
<br>
Dodatkowo pojawia się wartości rok takie jak: 
<ul>
    <li>
        <b>93, 95, 90, 80</b> - założenie jest takie, że w przypadku tych wartości jest to końcówka 19**, dlatego wstawiam <i>19</i> przed każdy z tych liczb
    </li>
    <li>
        <b>1</b> - za tę wartość wstawiam <i>np.NaN</i> (3 obiekty)
    </li>
    <li>
        <b>20521</b> - założenie, że miał być roku <i>2021</i> (rynek pierwotny)
    </li>
    <li>
        <b>19</b> - założenie, że miał być roku <i>2019</i>
    </li>
    <li>
        <b>202</b> - założenie, że miał być roku <i>2020</i> (rynek pierwotny)
    </li>
    <li>
        <b>1196</b> - za tę wartość wstawiam <i>np.NaN</i>
    </li>
</ul>

##### ENGLISH VERSION
<b>Rok budowy: 1930</b> <--- example of a Floor number value
<br>
<b>1930</b> <--- target appearance
<br>
Additionally there are exceptions like this:
<ul>
    <li>
        <b>93, 95, 90, 80</b> - I assume that these values are endings of 19**, imputed <i>19</i> before every single value
    </li>
    <li>
        <b>1</b> - in this case impute <i>np.NaN</i> (3 objects)
    </li>
    <li>
        <b>20521</b> - I assume that someone made a mistake, impute <i>2021</i> (rynek pierwotny)
    </li>
    <li>
        <b>19</b> - like above impute <i>2019</i>
    </li>
    <li>
        <b>202</b> - like above <i>2020</i> (rynek pierwotny)
    </li>
    <li>
        <b>1196</b> - in this case impute <i>np.NaN</i>
    </li>
</ul>

In [44]:
df['prod_year'] = df['prod_year'].apply(lambda x: pd.to_numeric(x.split(':')[1].strip()) if pd.notnull(x) else np.NaN)

In [45]:
df['prod_year'] = df['prod_year'].apply(lambda x: x+1900 if x == 93 or x == 95 or x == 90 or x == 80 else x)

In [46]:
df['prod_year'] = df['prod_year'].apply(lambda x: np.NaN if x == 1 else x)

In [47]:
df['prod_year'] = df['prod_year'].apply(lambda x: x+2000 if x == 19 else x)

In [48]:
df['prod_year'] = df['prod_year'].apply(lambda x: x+1818 if x == 202 else x)

In [49]:
df['prod_year'] = df['prod_year'].apply(lambda x: np.NaN if x == 1196 else x)

In [50]:
df['prod_year'] = df['prod_year'].apply(lambda x: x-18500 if x == 20521 else x)

In [51]:
df[df.prod_year >= 2023]

id   district     price  price_per_m2  floor  floor_number  \
1136  id48Fdc  Fabryczna  453177.0  6450.0        2.0    3.0            
2218  id48dOi  Krzyki     667275.0  7750.0        0.0    3.0            
7048  id48dSZ  Krzyki     448560.0  7200.0        1.0    2.0            
8174  id48dT1  Krzyki     426600.0  7900.0        0.0    2.0            

      prod_year            rent            market          building_type  \
1136  2031.0     Czynsz: 500 zł  Rynek: pierwotny  Rodzaj zabudowy: blok   
2218  2023.0     Czynsz: 750 zł  Rynek: pierwotny  Rodzaj zabudowy: blok   
7048  2023.0     Czynsz: 500 zł  Rynek: pierwotny  Rodzaj zabudowy: blok   
8174  2023.0     Czynsz: 450 zł  Rynek: pierwotny  Rodzaj zabudowy: blok   

                      finish_condition flat_surface_m2 room_number  
1136  Stan wykończenia: do wykończenia  70,26 m²        4 pokoje    
2218  Stan wykończenia: do wykończenia  86,10 m²        5 pokoi     
7048  Stan wykończenia: do wykończenia  62,30 m²        3 pokoje    
8174  Stan wykończenia: do wykończenia  54 m²           3 pokoje

In [52]:
df['prod_year'] = df['prod_year'].apply(lambda x: x - 10 if x == 2031 else x)

In [53]:
df.prod_year.unique()

array([1930., 2008., 2022., 2016., 2012., 2013., 2021., 2020.,   nan,
       2015., 2007., 1990., 2014., 2019., 1911., 1902., 2005., 2018.,
       1925., 2003., 2010., 1907., 1910., 1935., 1912., 1957., 1900.,
       2017., 1936., 2009., 2006., 1920., 1903., 1995., 2011., 1906.,
       2000., 1997., 1932., 2004., 1971., 1975., 1970., 1405., 2002.,
       1960., 1905., 1892., 1904., 1942., 1740., 1927., 1934., 1939.,
       1940., 1991., 2001., 1993., 1962., 1950., 1980., 1996., 1873.,
       1994., 1888., 1931., 1889., 1908., 1870., 1945., 1956., 1955.,
       1890., 1999., 1998., 1730., 1915., 1893., 1968., 1938., 1880.,
       1937., 1922., 1992., 1901., 1989., 1876., 1913., 1987., 1800.,
       1979., 1885., 2023., 1961., 1988., 1982., 1986., 1860., 1978.,
       1933., 1965., 1948., 1976., 1981., 1984., 1967., 1929., 1959.,
       1963., 1985., 1856., 1909., 1969., 1923., 1973., 1949., 1983.,
       1926., 1898., 1883., 1964., 1914., 1977., 1958., 1895., 1974.,
       1919., 1943.,

#### Rent

In [54]:
df.rent.head()

0    NaN           
1    Czynsz: 375 zł
2    Czynsz: 420 zł
3    NaN           
4    NaN           
Name: rent, dtype: object

In [55]:
(df.rent.unique())

array([nan, 'Czynsz: 375 zł', 'Czynsz: 420 zł', 'Czynsz: 2 800 zł',
       'Czynsz: 2 000 zł', 'Czynsz: 2 200 zł', 'Czynsz: 1 200 zł',
       'Czynsz: 2 300 zł', 'Czynsz: 7 zł', 'Czynsz: 1 800 zł',
       'Czynsz: 1 600 zł', 'Czynsz: 1 950 zł', 'Czynsz: 500 zł',
       'Czynsz: 850 zł', 'Czynsz: 800 zł', 'Czynsz: 1 100 zł',
       'Czynsz: 9 zł', 'Czynsz: 1 500 zł', 'Czynsz: 1 300 zł',
       'Czynsz: 1 207 zł', 'Czynsz: 400 zł', 'Czynsz: 1 350 zł',
       'Czynsz: 600 zł', 'Czynsz: 1 000 zł', 'Czynsz: 290 zł',
       'Czynsz: 900 zł', 'Czynsz: 12 zł', 'Czynsz: 1 416 zł',
       'Czynsz: 8 zł', 'Czynsz: 580 zł', 'Czynsz: 750 zł',
       'Czynsz: 300 zł', 'Czynsz: 913 zł', 'Czynsz: 1 020 zł',
       'Czynsz: 271 zł', 'Czynsz: 408 zł', 'Czynsz: 210 zł',
       'Czynsz: 700 zł', 'Czynsz: 1,10 zł', 'Czynsz: 270 zł',
       'Czynsz: 830 zł', 'Czynsz: 10 zł', 'Czynsz: 240 zł',
       'Czynsz: 380 zł', 'Czynsz: 680 zł', 'Czynsz: 480 zł',
       'Czynsz: 450 zł', 'Czynsz: 530 zł', 'Czynsz: 1 3

<b>Czynsz: 375 zł</b> <--- tak wygląda surowa wartość
<br>
<b>375.0</b> <--- docelowa wartość
<br>
Dodatkowo pojawia się wartości czynszu takie jak: 
<ul>
    <li>
        <b>9, 8, 1, 6, 7.4, 4, 12, 7, 1.1, 10, 0.1, 50, 90, 5.5, 70, 8.5, 3, 30, 2, 37, 42</b> - założenie jest takie, że w przypadku tych wartości wstawiam <i>np.NaN</i> ponieważ jest to oczywiście nie prawdziwe
    </li>
</ul>

##### ENGLISH VERSION
<b>Czynsz: 375 zł</b> <--- example of a rent value
<br>
<b>375.0</b> <--- target appearance
<br>
Additionally there are exceptions like this:
<ul>
    <li>
        <b>9, 8, 1, 6, 7.4, 4, 12, 7, 1.1, 10, 0.1, 50, 90, 5.5, 70, 8.5, 3, 30, 2, 37, 42</b> - in these cases impute <i>np.NaN</i> because it is impossible to have 8 PLN rent 
    </li>
</ul>

In [56]:
df['rent'] = df['rent'].apply(lambda x: pd.to_numeric(x.split(':')[1].split('z')[0].replace(' ','').replace(',','.').strip()) 
                              if pd.notnull(x) else np.NaN)

In [57]:
df.rent = df.rent.apply(lambda x: np.NaN if x in [9, 8, 1, 6, 7.4, 4, 12, 7, 1.1, 10, 0.1, 50, 90, 5.5, 70, 8.5, 3, 30, 
                                                  2, 37, 42 ] else x)

In [58]:
df.rent.unique()

array([    nan,  375.  ,  420.  , 2800.  , 2000.  , 2200.  , 1200.  ,
       2300.  , 1800.  , 1600.  , 1950.  ,  500.  ,  850.  ,  800.  ,
       1100.  , 1500.  , 1300.  , 1207.  ,  400.  , 1350.  ,  600.  ,
       1000.  ,  290.  ,  900.  , 1416.  ,  580.  ,  750.  ,  300.  ,
        913.  , 1020.  ,  271.  ,  408.  ,  210.  ,  700.  ,  270.  ,
        830.  ,  240.  ,  380.  ,  680.  ,  480.  ,  450.  ,  530.  ,
       1330.  ,  473.  ,  550.  ,  720.  ,  350.  , 1450.  ,  536.  ,
        200.  ,  780.  ,  260.  ,  130.  ,  531.  ,  710.  ,  328.  ,
        650.  ,  820.  ,  415.  , 1400.  ,  537.  ,  527.  ,  660.  ,
        613.  ,  460.  ,  424.  ,  295.  ,  199.  ,  206.  ,  390.  ,
        654.  ,  280.  ,  410.  ,  405.  ,  369.  ,  180.  ,  560.  ,
        440.  , 1220.  ,  370.  ,  250.  ,  190.  ,  230.  ,  745.  ,
        840.  ,  670.  ,  597.  ,  540.  , 1094.  ,  630.  ,  799.  ,
        760.  ,  490.  ,  907.  ,  475.  ,  747.  ,  510.  ,  555.  ,
        229.  ,  470

In [59]:
df.rent.head()

0   NaN   
1    375.0
2    420.0
3   NaN   
4   NaN   
Name: rent, dtype: float64

#### Market

In [60]:
df.market.head()

0    Rynek: wtórny   
1    Rynek: wtórny   
2    Rynek: pierwotny
3    Rynek: pierwotny
4    Rynek: wtórny   
Name: market, dtype: object

<b>Rynek: pierwotny</b> <--- tak wygląda surowa wartość
<br>
<b>pierwotny</b> <--- docelowa wartość
##### ENGLISH VERSION
<b>Rynek: pierwotny</b> <--- example of a market value
<br>
<b>pierwotny</b> <--- target appearance

In [61]:
df['market'] = df['market'].apply(lambda x: x.split(':')[1].strip() if pd.notnull(x) else np.NaN)

In [62]:
df.market.head()

0    wtórny   
1    wtórny   
2    pierwotny
3    pierwotny
4    wtórny   
Name: market, dtype: object

In [63]:
df.market.unique()

array(['wtórny', 'pierwotny'], dtype=object)

In [64]:
df.market.isnull().sum()

0

#### Building Type

In [65]:
df.building_type.head()

0    Rodzaj zabudowy: kamienica      
1    Rodzaj zabudowy: apartamentowiec
2    Rodzaj zabudowy: apartamentowiec
3    Rodzaj zabudowy: apartamentowiec
4    Rodzaj zabudowy: apartamentowiec
Name: building_type, dtype: object

<b>Rodzaj zabudowy: kamienica</b> <--- tak wygląda surowa wartość
<br>
<b>kamienica</b> <--- docelowa wartość
##### ENGLISH VERSION
<b>Rodzaj zabudowy: kamienica</b> <--- example of a building type value
<br>
<b>kamienica</b> <--- target appearance

In [66]:
df.building_type.unique()

array(['Rodzaj zabudowy: kamienica', 'Rodzaj zabudowy: apartamentowiec',
       'Rodzaj zabudowy: blok', nan, 'Rodzaj zabudowy: dom wolnostojący',
       'Rodzaj zabudowy: loft', 'Rodzaj zabudowy: szeregowiec',
       'Rodzaj zabudowy: plomba', 'Rodzaj zabudowy:'], dtype=object)

In [67]:
df['building_type'] = df['building_type'].apply(lambda x: x.split(':')[1].strip() if pd.notnull(x) else np.NaN)

In [68]:
df.building_type.unique()

array(['kamienica', 'apartamentowiec', 'blok', nan, 'dom wolnostojący',
       'loft', 'szeregowiec', 'plomba', ''], dtype=object)

#### Finish Condition

In [69]:
df.finish_condition.head()

0    NaN                              
1    Stan wykończenia: do zamieszkania
2    Stan wykończenia: do wykończenia 
3    Stan wykończenia: do zamieszkania
4    Stan wykończenia: do zamieszkania
Name: finish_condition, dtype: object

In [70]:
df.finish_condition.unique()

array([nan, 'Stan wykończenia: do zamieszkania',
       'Stan wykończenia: do wykończenia', 'Stan wykończenia: do remontu'],
      dtype=object)

<b>Stan wykończenia: do zamieszkania</b> <--- tak wygląda surowa wartość
<br>
<b>do zamieszkania</b> <--- docelowa wartość
##### ENGLISH VERSION
<b>Stan wykończenia: do zamieszkania</b> <--- example of a finish condition value
<br>
<b>do zamieszkania</b> <--- target appearance

In [71]:
df['finish_condition'] = df['finish_condition'].apply(lambda x: x.split(':')[1].strip() if pd.notnull(x) else np.NaN)

In [72]:
df.finish_condition.head()

0    NaN            
1    do zamieszkania
2    do wykończenia 
3    do zamieszkania
4    do zamieszkania
Name: finish_condition, dtype: object

In [73]:
df.finish_condition.unique()

array([nan, 'do zamieszkania', 'do wykończenia', 'do remontu'],
      dtype=object)

In [74]:
df.finish_condition.isnull().sum()

8553

#### Flat Surface m2

In [75]:
df.flat_surface_m2.head()

0    71,01 m² 
1    63 m²    
2    47,85 m² 
3    212,10 m²
4    214,79 m²
Name: flat_surface_m2, dtype: object

In [76]:
df.flat_surface_m2.unique()

array(['71,01 m²', '63 m²', '47,85 m²', ..., '38,38 m²', '46,91 m²',
       '49,15 m²'], dtype=object)

<b>71,01 m²</b> <--- tak wygląda surowa wartość
<br>
<b>71.01</b> <--- docelowa wartość
##### ENGLISH VERSION
<b>71,01 m²</b> <--- example of a flat surface in m2 value
<br>
<b>71.01</b> <--- target appearance

In [77]:
df['flat_surface_m2'] = df['flat_surface_m2'].apply(lambda x: pd.to_numeric(x.split('m')[0].replace(',','.').strip()) 
                                                    if pd.notnull(x) else np.NaN)

In [78]:
df.flat_surface_m2.head()

0    71.01 
1    63.00 
2    47.85 
3    212.10
4    214.79
Name: flat_surface_m2, dtype: float64

In [79]:
df.flat_surface_m2.unique()

array([71.01, 63.  , 47.85, ..., 38.38, 46.91, 49.15])

#### Room Number

In [80]:
df.room_number.head()

0    6 pokoi 
1    2 pokoje
2    2 pokoje
3    4 pokoje
4    4 pokoje
Name: room_number, dtype: object

<b>6 pokoi</b> <--- tak wygląda surowa wartość
<br>
<b>6</b> <--- docelowa wartość
<br>
Dodatkowo pojawia się wartości w liczbie pokoii takie jak: 
<ul>
    <li>
        <b>>10</b> - większa ilość pokoii niż 10 brak sprecyzowania 
    </li>
</ul>

##### ENGLISH VERSION
<b>6 pokoi</b> <--- example of a room number value
<br>
<b>6</b> <--- target appearance
<br>
Additionally there are exceptions like this: 
<ul>
    <li>
        <b>10</b> - in this case impute <i>np.NaN</i>
    </li>
</ul>

In [81]:
df.room_number.unique()

array(['6 pokoi', '2 pokoje', '4 pokoje', '5 pokoi', '3 pokoje',
       '7 pokoi', '10 pokoi', '9 pokoi', '1 pokój', '8 pokoi',
       '>10 pokoi'], dtype=object)

In [82]:
df[df['room_number'] == '>10 pokoi']

id   district      price  price_per_m2  floor  floor_number  \
127   id48CSc  Krzyki     1589500.0  15432.0      NaN     47.0           
7424  id48WFo  Fabryczna  440000.0   7333.0        1.0    2.0            

      prod_year    rent  market    building_type finish_condition  \
127   2013.0     1600.0  wtórny  apartamentowiec  do wykończenia    
7424 NaN        NaN      wtórny  NaN              NaN               

      flat_surface_m2 room_number  
127   103.0            >10 pokoi   
7424  60.0             >10 pokoi

In [83]:
df['room_number'] = df['room_number'].apply(lambda x: np.NaN if x == '>10 pokoi' else x)

In [84]:
df.room_number.isna().sum()

2

In [85]:
df['room_number'] = df['room_number'].apply(lambda x: pd.to_numeric(x.split(' ')[0].strip().replace('>10','6')) if pd.notnull(x)  else np.NaN)

In [86]:
df.room_number.head()

0    6.0
1    2.0
2    2.0
3    4.0
4    4.0
Name: room_number, dtype: float64

In [87]:
df.room_number.unique()

array([ 6.,  2.,  4.,  5.,  3.,  7., 10.,  9.,  1.,  8., nan])

In [88]:
df.isna().sum().sort_values(ascending=False)

rent                10266
finish_condition    8553 
building_type       5237 
prod_year           2512 
floor_number        2455 
district            704  
floor               456  
price               297  
price_per_m2        296  
room_number         2    
flat_surface_m2     0    
market              0    
id                  0    
dtype: int64

In [89]:
missing(df)

rent                68.28
finish_condition    56.88
building_type       34.83
prod_year           16.71
floor_number        16.33
district            4.68 
floor               3.03 
price               1.98 
price_per_m2        1.97 
room_number         0.01 
flat_surface_m2     0.00 
market              0.00 
id                  0.00 
dtype: float64


In [90]:
df.dtypes

id                  object 
district            object 
price               float64
price_per_m2        float64
floor               float64
floor_number        float64
prod_year           float64
rent                float64
market              object 
building_type       object 
finish_condition    object 
flat_surface_m2     float64
room_number         float64
dtype: object

In [91]:
df.describe()

price   price_per_m2         floor  floor_number     prod_year  \
count  1.473900e+04  14740.000000   14580.000000  12581.000000  12524.000000   
mean   5.117522e+05  8739.794030    2.343278      4.739130      2007.240897    
std    2.690295e+05  2295.186654    1.966537      3.186815      28.499904      
min    1.000000e+05  2308.000000   -1.000000      1.000000      1405.000000    
25%    3.790000e+05  7300.000000    1.000000      3.000000      2010.000000    
50%    4.500000e+05  8375.000000    2.000000      4.000000      2020.000000    
75%    5.650000e+05  9673.000000    3.000000      6.000000      2021.000000    
max    7.500000e+06  107370.000000  10.000000     52.000000     2023.000000    

              rent  flat_surface_m2   room_number  
count  4770.000000  15036.000000     15034.000000  
mean   467.079935   59.319652        2.723360      
std    216.653692   23.146364        0.961179      
min    7.500000     13.000000        1.000000      
25%    350.000000   45.472500        2.000000      
50%    449.500000   54.800000        3.000000      
75%    550.000000   69.000000        3.000000      
max    4500.000000  750.000000       10.000000

In [92]:
df.head(10)

id      district      price  price_per_m2  floor  floor_number  \
0  id48H25  Krzyki        590000.0   8309.0        3.0    3.0            
1  id48IAh  Krzyki        585000.0   9286.0        3.0    3.0            
2  id48Ihn  Stare Miasto  459000.0   9592.0        4.0    7.0            
3  id47fLE  Stare Miasto  7500000.0  35361.0       8.0    9.0            
4  id45AYc  Stare Miasto  7000000.0  32590.0       6.0    6.0            
5  id44eWl  Krzyki        5245000.0  23106.0      NaN     48.0           
6  id44gg6  Krzyki        4800000.0  27195.0      NaN     48.0           
7  id48Le4  Krzyki        4600000.0  34074.0      NaN     51.0           
8  id48jKt  Stare Miasto  4000000.0  29851.0       1.0    3.0            
9  id450d0  Stare Miasto  3582800.0  20000.0       9.0    10.0           

   prod_year    rent     market    building_type finish_condition  \
0  1930.0    NaN      wtórny     kamienica        NaN               
1  2008.0     375.0   wtórny     apartamentowiec  do zamieszkania   
2  2022.0     420.0   pierwotny  apartamentowiec  do wykończenia    
3  2016.0    NaN      pierwotny  apartamentowiec  do zamieszkania   
4  2016.0    NaN      wtórny     apartamentowiec  do zamieszkania   
5  2012.0     2800.0  pierwotny  apartamentowiec  do wykończenia    
6  2012.0     2000.0  pierwotny  apartamentowiec  do wykończenia    
7  2012.0    NaN      wtórny     apartamentowiec  do zamieszkania   
8  2013.0    NaN      wtórny     apartamentowiec  NaN               
9  2021.0     2200.0  pierwotny  apartamentowiec  do wykończenia    

   flat_surface_m2  room_number  
0  71.01            6.0          
1  63.00            2.0          
2  47.85            2.0          
3  212.10           4.0          
4  214.79           4.0          
5  227.00           5.0          
6  176.50           4.0          
7  135.00           3.0          
8  134.00           3.0          
9  179.14           6.0

In [93]:
df.to_csv('data_set_after_cleaning_with_NaN.csv',index=False)

# READ FILE WITH DATA AND NAN VALUES

In [94]:
df = pd.read_csv('data_set_after_cleaning_with_NaN.csv')

In [95]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15036 entries, 0 to 15035
Data columns (total 13 columns):
id                  15036 non-null object
district            14332 non-null object
price               14739 non-null float64
price_per_m2        14740 non-null float64
floor               14580 non-null float64
floor_number        12581 non-null float64
prod_year           12524 non-null float64
rent                4770 non-null float64
market              15036 non-null object
building_type       9797 non-null object
finish_condition    6483 non-null object
flat_surface_m2     15036 non-null float64
room_number         15034 non-null float64
dtypes: float64(8), object(5)
memory usage: 1.5+ MB


In [96]:
df.drop(columns=['id','rent', 'finish_condition'], inplace=True)

In [97]:
df = df[['district', 'building_type', 'market', 'room_number', 'floor_number', 'floor', 'prod_year', 'flat_surface_m2', 
             'price_per_m2', 'price']]

In [98]:
# df2 = df.query('market != "wtórny"').copy()
df = df[df['market'] != 'wtórny'].copy()

In [99]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9358 entries, 2 to 15035
Data columns (total 10 columns):
district           8998 non-null object
building_type      5019 non-null object
market             9358 non-null object
room_number        9358 non-null float64
floor_number       7243 non-null float64
floor              9086 non-null float64
prod_year          7994 non-null float64
flat_surface_m2    9358 non-null float64
price_per_m2       9064 non-null float64
price              9063 non-null float64
dtypes: float64(7), object(3)
memory usage: 804.2+ KB


In [100]:
df.isna().sum()

district           360 
building_type      4339
market             0   
room_number        0   
floor_number       2115
floor              272 
prod_year          1364
flat_surface_m2    0   
price_per_m2       294 
price              295 
dtype: int64

In [101]:
# GENERATES AND RETURN LIST OF INDEXES WITH NAN VALUE
def save_nan_index(column_name):
    index = df[column_name].index[df[column_name].apply(np.isnan)]
    df_index = df.index.values.tolist()
    index_list = [df_index.index(i) for i in index]
    return index_list

In [102]:
price_nan_index = save_nan_index('price')
price_per_m2_nan_index = save_nan_index('price_per_m2')
prod_year_nan_index = save_nan_index('prod_year')
floor_nan_index = save_nan_index('floor')
floor_number_nan_index = save_nan_index('floor_number')
# CATEGORICAL INDEX NO NEEDED BECAUSE I USED OTHER IMPUTE METHOD
# building_type_nan_index = save_nan_index('building_type')
# district_nan_index = save_nan_index('district')

In [103]:
df.head()

district    building_type     market  room_number  floor_number  floor  \
2  Stare Miasto  apartamentowiec  pierwotny  2.0          7.0           4.0     
3  Stare Miasto  apartamentowiec  pierwotny  4.0          9.0           8.0     
5  Krzyki        apartamentowiec  pierwotny  5.0          48.0         NaN      
6  Krzyki        apartamentowiec  pierwotny  4.0          48.0         NaN      
9  Stare Miasto  apartamentowiec  pierwotny  6.0          10.0          9.0     

   prod_year  flat_surface_m2  price_per_m2      price  
2  2022.0     47.85            9592.0        459000.0   
3  2016.0     212.10           35361.0       7500000.0  
5  2012.0     227.00           23106.0       5245000.0  
6  2012.0     176.50           27195.0       4800000.0  
9  2021.0     179.14           20000.0       3582800.0

In [104]:
from missingpy import MissForest
imputer = MissForest(random_state=100)
df_imputed = df.copy()
df_imputed = df_imputed.drop(columns=['district','building_type','market'])
df_imputed = imputer.fit_transform(df_imputed)

Iteration: 0
Iteration: 1
Iteration: 2
Iteration: 3
Iteration: 4
Iteration: 5
Iteration: 6


In [105]:
df_imputed = pd.DataFrame(data=df_imputed,columns=['room_number','floor_number','floor','prod_year','flat_surface_m2','price_per_m2','price'])
df_imputed = df_imputed.round(decimals=2)
df_imputed.floor_number = df_imputed.floor_number.round(decimals=0)
df_imputed.floor = df_imputed.floor.round(decimals=0)
df_imputed.prod_year = df_imputed.prod_year.round(decimals=0)

In [106]:
df_imputed.iloc[sorted([*prod_year_nan_index, *price_nan_index, *price_per_m2_nan_index, *floor_nan_index, *floor_number_nan_index])]

room_number  floor_number  floor  prod_year  flat_surface_m2  \
2     5.0          48.0          8.0    2012.0     227.00            
3     4.0          48.0          9.0    2012.0     176.50            
7     5.0          51.0          8.0    2012.0     183.80            
9     6.0          9.0           8.0    2021.0     179.33            
11    6.0          10.0          8.0    2021.0     180.00            
...   ...           ...          ...       ...        ...            
9349  3.0          4.0           1.0    2021.0     55.19             
9350  2.0          4.0           4.0    2020.0     47.70             
9352  2.0          4.0           3.0    2020.0     49.30             
9354  3.0          5.0           5.0    2021.0     47.80             
9355  2.0          4.0           1.0    2020.0     51.03             

      price_per_m2      price  
2     23106.0       5245000.0  
3     27195.0       4800000.0  
7     18904.0       3474555.0  
9     19000.0       3407270.0  
11    18929.0       3407270.0  
...       ...             ...  
9349  6570.0        362598.0   
9350  7600.0        362520.0   
9352  7350.0        362355.0   
9354  7575.0        362063.0   
9355  7094.0        362000.0   

[4340 rows x 7 columns]

## CATEGORICAL DATA IMPUTATION

In [107]:
df_imputed['district'] = df['district'].copy()

In [108]:
df_imputed['building_type'] = df['building_type'].copy()

In [109]:
df_imputed.district.mode()

0    Krzyki
dtype: object

In [110]:
df_imputed.district.fillna('Krzyki', inplace=True)

In [111]:
df_imputed.building_type.mode()

0    blok
dtype: object

In [112]:
df_imputed.building_type.fillna('blok', inplace=True)

In [113]:
df_imputed

room_number  floor_number  floor  prod_year  flat_surface_m2  \
0     2.0          7.0           4.0    2022.0     47.85             
1     4.0          9.0           8.0    2016.0     212.10            
2     5.0          48.0          8.0    2012.0     227.00            
3     4.0          48.0          9.0    2012.0     176.50            
4     6.0          10.0          9.0    2021.0     179.14            
...   ...           ...          ...       ...        ...            
9353  2.0          4.0           1.0    2020.0     51.03             
9354  3.0          5.0           5.0    2021.0     47.80             
9355  2.0          4.0           1.0    2020.0     51.03             
9356  3.0          2.0           0.0    2021.0     52.20             
9357  2.0          4.0           1.0    2021.0     38.20             

      price_per_m2      price      district    building_type  
0     9592.0        459000.0   Krzyki        blok             
1     35361.0       7500000.0  Krzyki        blok             
2     23106.0       5245000.0  Stare Miasto  apartamentowiec  
3     27195.0       4800000.0  Stare Miasto  apartamentowiec  
4     20000.0       3582800.0  Krzyki        blok             
...       ...             ...     ...         ...             
9353  7100.0        362313.0   Krzyki        blok             
9354  7575.0        362063.0   Krzyki        blok             
9355  7094.0        362000.0   Krzyki        blok             
9356  6933.0        361900.0   Psie Pole     blok             
9357  9424.0        360000.0   Psie Pole     blok             

[9358 rows x 9 columns]

In [114]:
df_imputed.floor = df_imputed.floor.astype('int64')
df_imputed.floor_number = df_imputed.floor_number.astype('int64')
df_imputed.prod_year = df_imputed.prod_year.astype('int64')
df_imputed.room_number = df_imputed.room_number.astype('int64')

In [115]:
df_imputed.describe().round(decimals=2)

room_number  floor_number    floor  prod_year  flat_surface_m2  \
count  9358.00      9358.00       9358.00  9358.00    9358.00           
mean   2.75         4.37          2.16     2020.16    58.67             
std    0.96         2.27          1.79     5.23       21.83             
min    1.00         1.00          0.00     1894.00    16.97             
25%    2.00         3.00          1.00     2020.00    44.90             
50%    3.00         4.00          2.00     2020.00    54.00             
75%    3.00         5.00          3.00     2021.00    67.87             
max    8.00         51.00         10.00    2023.00    305.82            

       price_per_m2       price  
count  9358.00       9358.00     
mean   8724.95       507907.30   
std    2077.47       270827.66   
min    3840.00       128944.00   
25%    7220.50       379000.00   
50%    8425.00       449000.00   
75%    9759.00       552716.50   
max    35361.00      7500000.00

In [116]:
df_imputed.to_csv('flats_wroclaw_final.csv',index=False)

## MISSFOREST ALGORITHM TEST

In [118]:
# Generate unique lists of random integers
inds1 = list(set(np.random.randint(0, len(df_without_nan), 10)))
inds2 = list(set(np.random.randint(0, len(df_without_nan), 15)))

# Replace the values at given index position with NaNs
# df_nan['price'] = [val if i not in inds1 else np.nan for i, val in enumerate(df_nan['price'])]
df_without_nan['room_number'] = [val if i not in inds1 else np.nan for i, val in enumerate(df_without_nan['room_number'])]
df_without_nan['floor'] = [val if i not in inds2 else np.nan for i, val in enumerate(df_without_nan['floor'])]
# df_nan['price_per_m2'] = [val if i not in inds2 else np.nan for i, val in enumerate(df_nan['price_per_m2'])]


# Get count of missing values by column
df_without_nan.isnull().sum()

In [ ]:
df_orygi = df_nan.dropna().copy()

In [ ]:
from missingpy import MissForest
imputer = MissForest(random_state=100)
X = df_nan_drop.drop(columns=['district','building_type','price','price_per_m2','flat_surface_m2','prod_year'],axis=1)
X_imputed = imputer.fit_transform(X)

In [ ]:
# df_nan['MF_price'] = X_imputed[:, 0]
df_orygi['MF_room_number'] = X_imputed[:, 0]
df_orygi['MF_floor'] = X_imputed[:, 2]
# df_nan['MF_price_per_m2'] = X_imputed[:, 6]
# comparison_df = df_nan[['price', 'MF_price', 'room_number', 'MF_room_number', 'floor','MF_floor','price_per_m2','MF_price_per_m2']]
comparison_df = df_orygi[['room_number', 'MF_room_number', 'floor','MF_floor']]
# comparison_df['ABS_ERROR_price'] = np.abs(comparison_df['price'] - comparison_df['MF_price'])
comparison_df['ABS_ERROR_room_number'] = np.abs(comparison_df['room_number'] - comparison_df['MF_room_number'])
comparison_df['ABS_ERROR_floor'] = np.abs(comparison_df['floor'] - comparison_df['MF_floor'])
# comparison_df['ABS_ERROR_price_per_m2'] = np.abs(comparison_df['price_per_m2'] - comparison_df['MF_price_per_m2'])
comparison_df.round(decimals=2)
comparison_df.iloc[sorted([*inds1, *inds2])].head(59)

In [ ]:
columns = list(df_nan.columns)
columns.remove('district')
columns.remove('building_type')

In [ ]:
df_imputed = pd.DataFrame(data=Xtransform,columns=columns)

#### This option displays full DataFrame

In [ ]:
pd.set_option('display.max_rows', df.shape[0]+1)